File: Contests/AtCoderPython/AtCoder Regular Contest 158/PB - Sum-Product Ratio.py
import sys

# import math
# from bisect import *
# from collections import *
# from functools import *
# from heapq import *
# from itertools import *
# from random import *
# from string import *
# from types import GeneratorType

# region fastio
input = lambda: sys.stdin.readline().rstrip()
sint = lambda: int(input())
mint = lambda: map(int, input().split())
ints = lambda: list(map(int, input().split()))
# print = lambda d: sys.stdout.write(str(d) + "\n")
# endregion fastio

def solve() -> None:
    n = sint()
    nums = ints()

    neg = [a for a in nums if a < 0]
    pos = [a for a in nums if a > 0]

    neg.sort(reverse = True)
    pos.sort(reverse = True)

    x = neg + pos

    def f(a, b, c):
        return (a + b + c) / (a * b * c)

    t1 = f(x[0], x[1], x[2])
    t2 = f(x[0], x[1], x[-1])
    t3 = f(x[0], x[-1], x[-2])
    t4 = f(x[-1], x[-2], x[-3])

    print(min(t1, t2, t3, t4))
    print(max(t1, t2, t3, t4))

solve()

File: Contests/AtCoderPython/AtCoder Regular Contest 158/PB - Sum-Product Ratio.go
package main

import (
	"bufio"
	. "fmt"
	"io"
	"math"
	"os"
	"sort"
)

func solve(_r io.Reader, _w io.Writer) {
	in := bufio.NewReader(_r)
	out := bufio.NewWriter(_w)
	defer out.Flush()

	var n, a int
	neg := []int{}
	pos := []int{}
	Fscan(in, &n)
	for i := 0; i < n; i++ {
		Fscan(in, &a)
		if a < 0 {
			neg = append(neg, a)
		} else {
			pos = append(pos, a)
		}
	}
	sort.Sort(sort.Reverse(sort.IntSlice(neg)))
	sort.Sort(sort.Reverse(sort.IntSlice(pos)))
	neg = append(neg, pos...)
	mn, mx := 3., -3.
	res := float64(neg[0]+neg[1]+neg[2]) / float64(neg[0]*neg[1]*neg[2])
	mn = math.Min(mn, res)
	mx = math.Max(mx, res)
	res = float64(neg[0]+neg[1]+neg[n-1]) / float64(neg[0]*neg[1]*neg[n-1])
	mn = math.Min(mn, res)
	mx = math.Max(mx, res)
	res = float64(neg[0]+neg[n-1]+neg[n-2]) / float64(neg[0]*neg[n-1]*neg[n-2])
	mn = math.Min(mn, res)
	mx = math.Max(mx, res)
	res = float64(neg[n-1]+neg[n-2]+neg[n-3]) / float64(neg[n-1]*neg[n-2]*neg[n-3])
	mn = math.Min(mn, res)
	mx = math.Max(mx, res)
	Fprintln(out, mn)
	Fprintln(out, mx)
}

func main() { solve(os.Stdin, os.Stdout) }
